src/server/worker/branch_set.go

```diff
@@ -51,81 +58,207 @@ func (a *APIServer) newBranchSetFactory(_ctx context.Context) (branchSetFactory,
-					rootCommits := make(map[string]map[string]bool)
-					setRootCommit := func(commit *pfs.Commit) {
-						if rootCommits[commit.Repo.Name] == nil {
-							rootCommits[commit.Repo.Name] = make(map[string]bool)
+					// Now we look for a set of commits that have provenance
+					// commits from the root input repos.  The set is only valid
+					// if there's precisely one provenance commit from each root
+					// input repo.
+					if set := findCommitSet(commitSets, i, func(set []*pfs.CommitInfo) bool {
+						rootCommits := make(map[string]map[string]bool)
+						setRootCommit := func(commit *pfs.Commit) {
+							if rootCommits[commit.Repo.Name] == nil {
+								rootCommits[commit.Repo.Name] = make(map[string]bool)
+							}
+							rootCommits[commit.Repo.Name][commit.ID] = true
 						}
-						rootCommits[commit.Repo.Name][commit.ID] = true
-					}
-					for _, commitInfo := range set {
-						setRootCommit(commitInfo.Commit)
-						for _, provCommit := range commitInfo.Provenance {
-							setRootCommit(provCommit)
+						for _, commitInfo := range set {
+							setRootCommit(commitInfo.Commit)
+							for _, provCommit := range commitInfo.Provenance {
+								setRootCommit(provCommit)
+							}
 						}
-					}
-					// ok tells us if it's ok to spawn a job with this
-					// commit set.
-					for _, rootInput := range rootInputs {
-						if len(rootCommits[rootInput.Repo]) != 1 {
-							return false
+						for _, rootInput := range rootInputs {
+							if rootInput.Atom != nil {
+								if len(rootCommits[rootInput.Atom.Repo]) != 1 {
+									return false
+								}
+							}
+						}
+						return true
+					}); set != nil {
+						bs := &branchSet{
+							NewBranch: i,
+						}
+						for i, commitInfo := range set {
+							branch := "master"
+							if directInputs[i].Atom != nil {
+								branch = directInputs[i].Atom.Branch
+							}
+							bs.Branches = append(bs.Branches, &pfs.BranchInfo{
+								Name: branch,
+								Head: commitInfo.Commit,
+							})
+						}
+						select {
+						case <-ctx.Done():
+							commitSetsMutex.Unlock()
+							return
+						case ch <- bs:
 						}
 					}
-					return true
-				}); set != nil {
-					bs := &branchSet{
-						NewBranch: i,
-					}
-					for i, commitInfo := range set {
-						bs.Branches = append(bs.Branches, &pfs.BranchInfo{
-							Name: directInputs[i].Branch,
-							Head: commitInfo.Commit,
-						})
-					}
-					select {
-					case <-ctx.Done():
+					commitSetsMutex.Unlock()
+				}
+			}()
+		}
+		if input.Cron != nil {
+			schedule, err := cron.Parse(input.Cron.Spec)
+			// it shouldn't be possible to error here because we validate the spec in CreatePipeline
+			if err != nil {
+				cancel()
+				return nil, err
+			}
+			tstamp := &types.Timestamp{}
+			var buffer bytes.Buffer
+			if err := pachClient.GetFile(input.Cron.Repo, "master", "time", 0, 0, &buffer); err != nil && !isNotFoundErr(err) {
+				cancel()
```

Comment:
I could be missing something, but instead of having all these `cancel()`s preceding `return`s, can we not just call `cancel` in a `defer` (if `retErr` is not `nil`)?